cEP-0024.md

```diff
@@ -0,0 +1,255 @@
+
+## Abstract
+
+This cEP describes the improvements to the coala workflow and the necessary
+automation in [GitMate](https://gitmate.io/) to realize it, as a part of the
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5344316227059712).
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there are a number of missing features which would save the time of
+coala community.
+
+Many of the plugins in GitMate are general plugins, but coala needs more
+plugins which support coala development workflow. So, this project
+adds plugins to auto-review, auto-reject, auto-label PRs, assign, unassign
+issues, squash commits.
+
+## Changes in coala workflow
+
+### Use increasing difficulty level order to assign issues
+
+The current assign process in coala is to check team membership of a user,
+difficulty level of the issue and assign the issue to user if eligible. To
+implement the current assign process in GitMate, IGitt would need to support
+teams.
+
+Currently, IGitt doesn't support teams and there are no teams in GitLab.
+Considering [GitLab groups](https://docs.gitlab.com/ee/user/group/) and
+[GitHub teams](https://help.github.com/articles/about-teams/) as teams in
+IGitt is a possible solution for IGitt implementation. But, it requires extra
+tools to sync GitHub team and GitLab group memberships, which is a manual
+process and isn't automated yet.
+
+Also, team membership information is not publicly accessible in GitHub, so any
+system which needs to implement the current assign system would require
+organization authentication tokens.
+
+The best possible solution would be using an increasing difficulty level order,
+so that a user can be assigned to an issue only if the user has solved issues
+of lower difficulty level before requesting to get assigned to a higher
+difficulty level issue.
+
+#### Drawbacks
+
+If a user completed the lower difficulty level issues in GitHub and want to
+work on a higher difficulty level issue in GitLab, he/she has to repeat the
+newcomer process in GitLab. To resolve this issue, GitMate needs to find a way
+to link GitHub org and GitLab org, this would be easy when they have the same
+name, as in the case of `coala`. But, resolving this issue is out of scope for
+this project.
+
+If a user is worthy enough to work on the higher level difficulty issues
+directly, then that user needs to be manually assigned by the maintainers until
+that user completes at least one low difficulty issue. Even with current assign
+system, manual intervention by maintainers is needed to add a user to
+developers team.
+
+## New Features
+
+### Auto-assign issue when it is mentioned in any of PR commits
+
+When a PR is opened or updated, if any of PR's commit message referenced
+any issues with keywords `Fixes` or `Closes`, there is no
+CI failure in PR, then auto assign the author of the PR to the mentioned
```

Comment:
Will it auto-assign the author of the PR to the mentioned unassigned issues even if they are ineligible for the difficulty of the issues?

What happens if the issue has no difficulty level set yet?

Reply:
>Will it auto-assign the author of the PR to the mentioned unassigned issues even if they are ineligible for the difficulty of the issues?

No

>What happens if the issue has no difficulty level set yet?

GitMate will assign the PR author to that issue..

Reply:
I have explained about the eligibility in the next line:

>If the author of the PR isn't eligible to get assigned to
the mentioned issues, missing eligibility conditions will be specified as a
comment, which isn't configurable.